### PR TITLE
Micro web server

### PR DESCRIPTION
**Pull request by MartinWildbaum**

### CODE REVIEW

File: micro_web_server/views/index.erb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Exercise one Sinatra</title>
+    <link rel="stylesheet" type="text/css" href="/styles/styles.css">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.16.0/umd/popper.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js"></script>
+  </head>
+  <body>
+    <div class="container-fluid bg-custom-skyblue py-5 px-5">
+      <div class="conatiner d-flex flex-column gap-5 text-white">
```
Comment (cristianv77) [inline micro_web_server/views/index.erb]: typo
```suggestion
      <div class="container d-flex flex-column gap-5 text-white">
```

File: micro_web_server/server.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+post '/optimize' do
+  budget = params[:budget].to_f
+
+  if budget >= 0
+    $budget = budget
+    sort_items_by_cost
+    check_affordable_items
+  end
+
+  redirect'/'
```
Comment (cristianv77) [inline micro_web_server/server.rb]: space before '/'
```suggestion
  redirect '/'
```

File: micro_web_server/server.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+get '/' do
+  erb :index
+end
+
+post '/add_item' do
+  description = params[:description]
+  cost = params[:cost].to_f
+
+  if !description.empty? && cost >= 0
```
Comment (cristianv77) [inline micro_web_server/server.rb:39]: you can use description.present? instead

File: micro_web_server/server.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    item.save
+  end
+  $items = WishlistItem.all
+  redirect '/'
+end
+
+post '/optimize' do
+  budget = params[:budget].to_f
+
+  if budget >= 0
```
Comment (nachodiba210) [inline micro_web_server/server.rb:49]: do we want to sort items and everything if the budget params is nil? because it's happening, if we do nil.to_f it will give 0.0

File: micro_web_server/wishlist_item.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      puts "Item with name #{@name} already exists in the database."
+    end
+  end
+
+  def self.all
+    result = DB.exec('SELECT * FROM wishlist_items')
+    result.map do |row|
+      WishlistItem.new(row['name'], row['cost'].to_f, row['completed'] == 't')
+    end
+  end
```
Comment (nachodiba210) [inline micro_web_server/wishlist_item.rb:27]: why are you doing this methods? they both already exists

File: micro_web_server/wishlist_item.rb
Comment (nachodiba210) [inline micro_web_server/wishlist_item.rb:1]: i believe you are missing some key information about this part, the onboarding doesn't say it the best way, but for this part you will need to have a class for users also, we want to store their favorite items. so you will be able to login as a user, and create the list of their favorite items and optimize if needed, also be able to go into other user and have other favorite list.
We can discuss this on the daily

**Review by nachodiba210 — CHANGES_REQUESTED** [review]
I think you are missing the users and some parts of this exercise we can discuss on the daily

File: micro_web_server/server.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+require 'pg'
+
+DB = PG.connect(
+  dbname: 'sinatra_web_page',
+  user: 'martinjavierwildbaum',
+  password: 'PedroBerro819'
+)
+
+set :public_folder, File.dirname(__FILE__) + '/public'
+set :views, File.join(File.dirname(__FILE__), 'views')
```
Comment (nachodiba210) [inline micro_web_server/server.rb]: why is this needed?

File: micro_web_server/server.rb (hunk 4 above)
Comment (MartinWildbaum) [reply micro_web_server/server.rb:49]: I thought it was already controlled as it is not possible to enter a number and it is obligatory to enter something in order to add an item.
image
image
In fact, the verification regarding description  (if !description.empty? && cost >= 0) was done only to learn as it is already controlled. 
image
image

File: micro_web_server/server.rb (hunk 3 above)
Comment (MartinWildbaum) [reply micro_web_server/server.rb:39]: I believe that in order to use the method .present? I need an extra gem, while with .empty I do not

Comment (cristianv77) [reply micro_web_server/server.rb:39]: What about any?

File: micro_web_server/server.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
+require 'sinatra'
+require 'dotenv/load'
+require_relative 'wishlist_item'
+require 'pg'
+
+DB = PG.connect(
+  dbname: ENV['DB_NAME'],
+  user: ENV['USER'],
+  password: ENV['DB_PASS']
```
Comment (nachodiba210) [inline micro_web_server/server.rb:9]: good one!

File: micro_web_server/server.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+end
+
+get '/search' do
+  search_query = params[:text_to_filter]
+  if search_query
+    # Decided not to make it case sensitive.
+    @filtered_items = $items.select { |item| item.name.downcase.include?(search_query.downcase) }
+  else
+    @filtered_items = $items
+  end
```
Comment (nachodiba210) [inline micro_web_server/server.rb]:
```suggestion
 @filtered_items = if search_query
    # Decided not to make it case sensitive.
    $items.select { |item| item.name.downcase.include?(search_query.downcase) }
  else
     $items
  end
```

**Review by nachodiba210 — APPROVED** [review]
Good work!
